### PR TITLE
PRT-891 Reward Persistency - ft-agilefreaks | ft-distributed-lab

### PR DESCRIPTION
**Pull request by shleikes**
This PR adds the Badger DB as a snapshot database for the rewards server, to save the rewards that hadn't been confirmed yet, so in the rare case that the provider or the rewards server fail, or restarted, the rewards will be restored from the filesystem and the providers wont lose their rewards.

### CODE REVIEW

File: protocol/rpcprovider/rewardserver/reward_server.go [hunk 1]
```diff
@@ -272,23 +354,136 @@ func (rws *RewardServer) PaymentHandler(payment *PaymentRequest) {
...
+			rws.resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer()
+		case <-rws.rewardsSnapshotThresholdCh:
+			rws.resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer()
+		}
+	}
+}
+
+func (rws *RewardServer) resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer() {
+	// We lock without defer because the DB is already locking itself
+	rws.lock.RLock()
```
Comment (ranlavanet) [inline protocol/rpcprovider/rewardserver/reward_server.go:402]: rws.rewardsSnapshotTimer.Reset needs write privileges so not rlock -> lock.

File: protocol/rpcprovider/rewardserver/reward_server.go [hunk 2]
```diff
@@ -272,23 +354,136 @@ func (rws *RewardServer) PaymentHandler(payment *PaymentRequest) {
...
+		return
+	}
+
+	utils.LavaFormatDebug("saving rewards snapshot to the DB", utils.Attribute{Key: "proofs", Value: len(rewardEntities)})
+
+	var err error
+	for i := 0; i < MaxDBSaveRetries; i++ {
+		err = rws.rewardDB.BatchSave(rewardEntities)
+		if err == nil {
+			return
```
Comment (ranlavanet) [inline protocol/rpcprovider/rewardserver/reward_server.go:431]: let's add a success print INFO level

File: protocol/rpcprovider/rewardserver/reward_server.go [hunk 3]
```diff
@@ -272,23 +354,136 @@ func (rws *RewardServer) PaymentHandler(payment *PaymentRequest) {
...
+		}
+
+		utils.LavaFormatWarning("failed saving proofs snapshot to rewardDB. Retrying...", err,
```
Comment (ranlavanet) [inline protocol/rpcprovider/rewardserver/reward_server.go:434]: lets change it to INFO level instead of warning.

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/rpcprovider/rewardserver/reward_server.go (hunk 3 above)
Comment (shleikes) [reply protocol/rpcprovider/rewardserver/reward_server.go:430]: What about the error?

File: protocol/rpcprovider/rewardserver/badger_db.go [hunk 4]
```diff
@@ -16,7 +16,7 @@ type BadgerDB struct {
 	shardString  string
 	rewards      map[string]*entryWithTtl
 	db           *badger.DB
-	lock         sync.RWMutex
+	lock         sync.Mutex
```
Comment (ranlavanet) [inline protocol/rpcprovider/rewardserver/badger_db.go:19]: why change to Mutex? we can still Rlock for read and lock for write, it will be faster

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/rpcprovider/rewardserver/badger_db.go (hunk 4 above)
Comment (shleikes) [reply protocol/rpcprovider/rewardserver/badger_db.go:19]: After a second look, you're right, I'll change that backl

Comment (shleikes) [reply protocol/rpcprovider/rewardserver/badger_db.go]: Fixed

**Review by ranlavanet — APPROVED** [review]

**Pull request merged**